# Simple Linear Regression End-to-End
Predict `exam_score` from `study_hours` using manual OLS, scikit-learn, and statsmodels.

In [1]:
from pathlib import Path
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import statsmodels.api as sm

DATA_DIR = Path.cwd().parent / 'data' if (Path.cwd().parent / 'data').exists() else Path.cwd() / 'data'
df = pd.read_csv(DATA_DIR / 'simple_linear_student_scores.csv')
X = df[['study_hours']]
y = df['exam_score']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=43)
model = LinearRegression().fit(X_train, y_train)
pred = model.predict(X_test)
mae = mean_absolute_error(y_test, pred)
rmse = np.sqrt(mean_squared_error(y_test, pred))
r2 = r2_score(y_test, pred)
x = X_train['study_hours'].to_numpy()
yt = y_train.to_numpy()
manual_slope = ((x - x.mean()) * (yt - yt.mean())).sum() / ((x - x.mean()) ** 2).sum()
manual_intercept = yt.mean() - manual_slope * x.mean()
print(f'rows: {len(df)}')
print(f'intercept: {model.intercept_:.4f}')
print(f'slope: {model.coef_[0]:.4f}')
print(f'MAE: {mae:.4f}')
print(f'RMSE: {rmse:.4f}')
print(f'R2: {r2:.4f}')
print(f'manual slope: {manual_slope:.4f}')
print(f'manual intercept: {manual_intercept:.4f}')

rows: 60
intercept: 39.0182
slope: 5.4579
MAE: 4.7378
RMSE: 6.1580
R2: 0.7670
manual slope: 5.4579
manual intercept: 39.0182


In [2]:
ols = sm.OLS(y_train, sm.add_constant(X_train)).fit()
pd.DataFrame({
    'coef': ols.params,
    'p_value': ols.pvalues,
    'lower_95': ols.conf_int()[0],
    'upper_95': ols.conf_int()[1]
})

                  coef       p_value   lower_95   upper_95
const        39.018170  5.496479e-29  36.167446  41.868895
study_hours   5.457874  3.132511e-26   4.991270   5.924477

## Interpretation
The slope is approximately 5.46. In this synthetic dataset, one additional study hour is associated with about 5.46 additional score points. Students should compare this with `attendance_pct` as a one-feature model.